cmake/modules/hunter_download_server_url.cmake

```diff
@@ -0,0 +1,84 @@
+# Copyright (c) 2017 NeroBurner
+# All rights reserved.
+
+include(CMakeParseArguments) # cmake_parse_arguments
+
+include(hunter_status_print)
+include(hunter_internal_error)
+include(hunter_test_string_not_empty)
+
+# used gobal variables
+# - HUNTER_DOWNLOAD_SERVER (mandatory)
+#   - URLs are updated to point to the download server
+# - HUNTER_DOWNLOAD_SERVER_BLACKLIST
+#   - blacklist specifies server urls, which are not mapped to the download server
+
+# update download url to point to HUNTER_DOWNLOAD_SERVER
+# - PACKAGE:       name of the package to update the url for
+# - version:       version of the package
+# - URL:           original URL for the package
+# - OUTPUT:        variable to output the updated URL
+function(hunter_download_server_url)
+  set(function_name "hunter_download_server_url")
+  set(function_synopsis "${function_name}(PACKAGE foo VERSION 0.1.2-p3 URL http://original_url.example.com OUTPUT variable)")
+
+  # parse arguments
+  set(one_value_args PACKAGE VERSION URL OUTPUT)
+  cmake_parse_arguments(x "" "${one_value_args}" "" ${ARGV})
+
+  # No free arguments allowed
+  list(LENGTH x_UNPARSED_ARGUMENTS x_len)
+  if(NOT x_len EQUAL 0)
+    hunter_internal_error(
+      "'${function_name}' incorrect usage,"
+      " expected no free arguments '${x_UNPARSED_ARGUMENTS}'."
+      " Synopsis: ${function_synopsis}"
+    )
+  endif()
+
+  # check mandatory arguments
+  foreach(arg PACKAGE VERSION URL OUTPUT)
+    string(COMPARE EQUAL "${x_${arg}}" "" is_empty)
+    if(is_empty)
+      hunter_internal_error(
+        "'${function_name}' incorrect usage,"
+        " option '${arg}' with one argument is mandatory."
+        " Synopsis: ${function_synopsis}"
+      )
+    endif()
+  endforeach()
+
+  hunter_test_string_not_empty("${HUNTER_DOWNLOAD_SERVER}")
+
+  # update download url to point to HUNTER_DOWNLOAD_SERVER
+  # - blacklist specifies server urls, which are not mapped to the download server
+  set(pkg_url_in_blacklist FALSE)
+  foreach(blacklist_item ${HUNTER_DOWNLOAD_SERVER_BLACKLIST})
+    string(FIND "${x_URL}" "${blacklist_item}" found)
+    if(NOT (found EQUAL -1))
+      hunter_status_print("DOWNLOAD_SERVER: \"${x_PACKAGE}\": URL \"${x_URL}\" is blacklisted by \"${blacklist_item}\".")
+      set(pkg_url_in_blacklist TRUE)
+      break()
+    endif()
+  endforeach()
+
+  set(hunter_package_new_url "${x_URL}")
+  if (NOT pkg_url_in_blacklist)
+    # point download URL to DOWNLOAD_SERVER
+    set(hunter_package_original_url "${x_URL}" )
+    get_filename_component(package_filename_raw "${x_URL}" NAME)
+    # replace special characters from filename
+    string(REGEX REPLACE "[!@#$%^&*?]" "_" package_filename "${package_filename_raw}")
+    # create new URL
+    set(hunter_package_new_url "${HUNTER_DOWNLOAD_SERVER}/${x_PACKAGE}/${x_VERSION}/${package_filename}")
```

Comment:
May be it make sense to use this layout:
* https://docs.hunter.sh/en/latest/reference/layouts/deployed.html#download

It will be possible to share `~/.hunter/_Base/Download` as far as I understand. E.g. in terms of Nginx set `~/.hunter/_Base/Download` as a root location.

Reply:
where is archive-ID calculated? I think it was the first 7 digits of the sha1 of the package archive?

Reply:
> I think it was the first 7 digits of the sha1 of the package archive?

Yes, see https://docs.hunter.sh/en/latest/reference/layouts/deployed.html#common